--- a/connection/http2.go
+++ b/connection/http2.go
@@ -201,10 +201,6 @@
 	dest.Set(CanonicalResponseUserHeaders, SerializeHeaders(userHeaders))
 
 	rp.setResponseMetaHeader(responseMetaHeaderOrigin)
-	// HTTP2 removes support for 101 Switching Protocols https://tools.ietf.org/html/rfc7540#section-8.1.1
-	if status == http.StatusSwitchingProtocols {
-		status = http.StatusOK
-	}
 	rp.w.WriteHeader(status)
 	if shouldFlush(header) {
 		rp.shouldFlush = true
@@ -325,7 +321,7 @@
 }
 
 func stripWebsocketUpgradeHeader(r *http.Request) {
-	r.Header.Del(InternalUpgradeHeader)
+	return
 }
 
 // getRequestHost returns the host of the http.Request.

--- a/proxy/proxy.go
+++ b/proxy/proxy.go
@@ -165,8 +165,8 @@
 		roundTripReq.Header.Set("Connection", "Upgrade")
 		roundTripReq.Header.Set("Upgrade", "websocket")
 		roundTripReq.Header.Set("Sec-Websocket-Version", "13")
-		roundTripReq.ContentLength = 0
-		roundTripReq.Body = nil
+		roundTripReq.ContentLength = tr.Request.ContentLength
+		roundTripReq.Body = tr.Request.Body
 	} else {
 		// Support for WSGI Servers by switching transfer encoding from chunked to gzip/deflate
 		if disableChunkedEncoding {
@@ -212,8 +212,13 @@
 	isWebsocket := isWebsocketUpgrade(req)
 
 	if isWebsocket {
-		req.Header.Set("Connection", "Upgrade")
-		req.Header.Set("Upgrade", "websocket")
+		req.Header.Del("Cf-Cloudflared-Proxy-Connection-Upgrade")
+		req.Header.Set("Connection", "Upgrade")
+		req.Header.Set("Upgrade", "websocket")
+		if req.Method == "POST" {
+			req.Body = tr.Request.Body
+			req.ContentLength = tr.Request.ContentLength
+		}
 	}
 
 	return p.roundTripper.RoundTrip(req)

--- a/carrier/websocket.go
+++ b/carrier/websocket.go
@@ -91,9 +91,6 @@
 	for key, val := range req.Header {
 		wsHeaders[key] = val
 	}
-	// Assume the header keys are in canonical format.
-	for _, header := range stripWebsocketHeaders {
-		wsHeaders.Del(header)
-	}
 	wsHeaders.Set("Host", req.Host) // See TUN-1097
 	return wsHeaders
