--- a/connection/http2.go
+++ b/connection/http2.go
@@ -112,3 +112,3 @@
 	case TypeWebsocket, TypeHTTP:
-		stripWebsocketUpgradeHeader(r)
+		// [Headscale Fix]
 		// Check for tracing on request
@@ -328,3 +329,3 @@
 func stripWebsocketUpgradeHeader(r *http.Request) {
-	r.Header.Del(InternalUpgradeHeader)
+	// [Headscale Fix]
 }

--- a/proxy/proxy.go
+++ b/proxy/proxy.go
@@ -167,4 +167,9 @@
 		roundTripReq.Header.Set("Sec-Websocket-Version", "13")
-		roundTripReq.ContentLength = 0
-		roundTripReq.Body = nil
+		if tr.Request.Method == http.MethodGet || tr.Request.Method == http.MethodHead {
+			roundTripReq.ContentLength = 0
+			roundTripReq.Body = nil
+		} else {
+			roundTripReq.ContentLength = tr.Request.ContentLength
+			roundTripReq.Body = tr.Request.Body
+		}
 	} else {

--- a/carrier/websocket.go
+++ b/carrier/websocket.go
@@ -93,3 +93,3 @@
 	for _, header := range stripWebsocketHeaders {
-		wsHeaders.Del(header)
+		// [Headscale Fix]
 	}
