diff --git a/connection/http2.go b/connection/http2.go
index c7e14c67..1fc3885e 100644
--- a/connection/http2.go
+++ b/connection/http2.go
@@ -100,6 +100,21 @@ func (c *HTTP2Connection) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	c.activeRequestsWG.Add(1)
 	defer c.activeRequestsWG.Done()
 
+	fmt.Println("========== DUMP REQUEST ==========")
+	fmt.Printf("Method: %s | URL: %s | RemoteAddr: %s\n", r.Method, r.URL.String(), r.RemoteAddr)
+	fmt.Println("--- Headers ---")
+	for k, v := range r.Header {
+		fmt.Printf("%s: %v\n", k, v)
+	}
+	if r.Body != nil {
+			fmt.Printf("--- Body Info ---\n")
+			fmt.Printf("ContentLength: %d (Note: -1 means chunked/unknown)\n", r.ContentLength)
+			fmt.Printf("Body Object: %p (Non-nil)\n", r.Body)
+	} else {
+			fmt.Println("--- Body: nil ---")
+	}
+	fmt.Println("==================================")
+
 	connType := determineHTTP2Type(r)
 	handleMissingRequestParts(connType, r)
 
