diff --git a/proxy/proxy.go b/proxy/proxy.go
index e5d7fc6d..feca6629 100644
--- a/proxy/proxy.go
+++ b/proxy/proxy.go
@@ -197,8 +197,13 @@ func (p *Proxy) proxyHTTPRequest(
 		roundTripReq.Header.Set("Connection", "Upgrade")
 		roundTripReq.Header.Set("Upgrade", "websocket")
 		roundTripReq.Header.Set("Sec-Websocket-Version", "13")
-		roundTripReq.ContentLength = 0
-		roundTripReq.Body = nil
+		if tr.Request.Body != nil && tr.Request.ContentLength != 0 {
+			roundTripReq.ContentLength = tr.Request.ContentLength
+			roundTripReq.Body = tr.Request.Body
+		} else {
+			roundTripReq.ContentLength = 0
+			roundTripReq.Body = nil
+		}
 	} else {
 		// Support for WSGI Servers by switching transfer encoding from chunked to gzip/deflate
 		if disableChunkedEncoding {
