--- a/connection/http2.go
+++ b/connection/http2.go
@@ -325,7 +325,7 @@
 }
 
 func stripWebsocketUpgradeHeader(r *http.Request) {
-	r.Header.Del(InternalUpgradeHeader)
+	return 
 }
 
 // getRequestHost returns the host of the http.Request.

--- a/proxy/proxy.go
+++ b/proxy/proxy.go
@@ -165,8 +165,8 @@
 		roundTripReq.Header.Set("Connection", "Upgrade")
 		roundTripReq.Header.Set("Upgrade", "websocket")
 		roundTripReq.Header.Set("Sec-Websocket-Version", "13")
-		roundTripReq.ContentLength = 0
-		roundTripReq.Body = nil
+		roundTripReq.ContentLength = tr.Request.ContentLength
+		roundTripReq.Body = tr.Request.Body
 	} else {
 		// Support for WSGI Servers by switching transfer encoding from chunked to gzip/deflate
 		if disableChunkedEncoding {

--- a/carrier/websocket.go
+++ b/carrier/websocket.go
@@ -91,9 +91,6 @@
 	}
 	// Assume the header keys are in canonical format.
-	for _, header := range stripWebsocketHeaders {
-		wsHeaders.Del(header)
-	}
 	wsHeaders.Set("Host", req.Host) // See TUN-1097
 	return wsHeaders
